### PR TITLE
Implemented Reciprocal Rank Fusion

### PR DESCRIPTION
**Pull request by aarya-16**
# What does this PR do?

Implements Reciprocal Rank Fusion Algorithm using Ensemble Retriever from langchain to combine multiple retriever outputs, which addresses Issue #55

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Was this discussed/approved via a Github issue? Please add a link to it if that's the case
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag  members/contributors who
may be interested in your PR.

### CODE REVIEW

**Comment by aarya-16** [comment]
@iuliaturc @mihail911 can someone take a look at this PR and tell me if any changes are to be done ?

**Comment by iuliaturc** [comment]
Looking now! Sorry for the delay! I'd like to rerun [the benchmarks](https://github.com/Storia-AI/sage/tree/main/benchmarks/retrieval) to see if reciprocal rank fusion makes BM25 any better.

**Comment by iuliaturc** [comment]
Haven't forgotten about this, will make a final review tomorrow.

File: sage/vector_store.py [hunk 1]
```diff
@@ -133,19 +134,23 @@ def upsert_batch(self, vectors: List[Vector], namespace: str):
...
+        ) if self.bm25_encoder else None
+
+        dense_retriever = LangChainPinecone.from_existing_index(
             index_name=self.index_name, embedding=embeddings, namespace=namespace
         ).as_retriever(search_kwargs={"k": top_k})
+        
+        retrievers = [bm25_retriever, dense_retriever] if bm25_retriever else [dense_retriever]
+        if len(retrievers) == 1:
+            return retrievers[0]
+        return EnsembleRetriever(retrievers=retrievers, weights=[1.0 / len(retrievers)] * len(retrievers))
```
Comment (iuliaturc) [inline sage/vector_store.py:152]: Could you use self.alpha instead of equal weights so that we can weigh dense vs sparse?

**Comment by aarya-16** [comment]
@iuliaturc Changed it to use self.alpha

File: sage/vector_store.py [hunk 2]
```diff
@@ -133,19 +134,25 @@ def upsert_batch(self, vectors: List[Vector], namespace: str):
...
             index_name=self.index_name, embedding=embeddings, namespace=namespace
         ).as_retriever(search_kwargs={"k": top_k})
+        
+        retrievers = [bm25_retriever, dense_retriever] if bm25_retriever else [dense_retriever]
+        if len(retrievers) == 1:
+            return retrievers[0]
+        
+        weights = [1 - self.alpha, self.alpha] if bm25_retriever else [1.0]
+        return EnsembleRetriever(retrievers=retrievers, weights=weights)
+
```
[the remark this answers is not in the document]
Comment (aarya-16) [reply sage/vector_store.py]: Yeah I agree.

**Review by iuliaturc — APPROVED** [review]

**Pull request merged**